### PR TITLE
Add configuration to show all test messages.

### LINKED ISSUE #184826: `TestRun` `appendOutput` is lost when running an unrelated test.
cc @connor4312 In addition to <https://github.com/microsoft/vscode/issues/184825> I have another issue.

When running an unrelated test, the text provided to `appendOutput` can flash in and out of existence, or disappear entirely.

https://github.com/microsoft/vscode/assets/30030/cc04c107-cc53-4b45-ae9d-b15d76e312a7

I'm not sure why this is happening - I don't believe there is anything wrong with my code. My expectation is that after using `appendOutput`, the output should remain attached to that test until the test itself is re-run.

### PR DESCRIPTION
**Pull request by ioquatix**
Introduce `testing.showAllMessages` configuration, which defaults to true. When false, only the first message is shown, and the rest are ignored. When true, the most recent message for each line is shown from all past test runs.

Fixes #184826.

cc @connor4312

### CODE REVIEW

**Comment by ioquatix** [comment]
@microsoft-github-policy-service agree

File: src/vs/workbench/contrib/testing/browser/testingDecorations.ts [hunk 1]
```diff
@@ -318,47 +318,11 @@ export class TestingDecorationService extends Disposable implements ITestingDeco
...
-								resultId: lastResult.id,
-								testExtId: test.item.extId,
-							}), model);
-
-							newDecorations.addMessage(decoration);
-							messageLines.set(line, decoration.testMessage);
-						}
-					}
-				}
+			const messageLines = new Map</* line number */ number, /* last test message */ ITestMessage>();
```
Comment (connor4312) [inline src/vs/workbench/contrib/testing/browser/testingDecorations.ts:321]: This could now just be a `Set<number>` since the value is no longer used

**Review by connor4312 — COMMENTED** [review]
Looks good overall, one comment for a small optimization

**Review by precha1989 — APPROVED** [review]

File: src/vs/workbench/contrib/testing/browser/testingDecorations.ts (hunk 1 above)
Comment (ioquatix) [reply src/vs/workbench/contrib/testing/browser/testingDecorations.ts:321]: Good idea, I'll update it.e

**Comment by ioquatix** [comment]
I have a question around caching messages/decorations.

It would make sense to me to reuse/cache these decorations.

I see `CachedDecorations` 'lastDecorations` and `newDecorations`.

In the case this setting is true, can we avoid having to iterate through all results and just copy `lastDecorations` to `newDecorations` (or just reuse `lastDecorations` directly), to avoid having to iterate all the old test runs?

If so, can you provide any suggestions about the best way to implement that, and I'll update the PR.

**Comment by connor4312** [comment]
It looks like there's a compilation error

> In the case this setting is true, can we avoid having to iterate through all results and just copy lastDecorations to newDecorations (or just reuse lastDecorations directly), to avoid having to iterate all the old test runs?

We could do that. I would probably add a new property on CachedDecorations that maps a TestResult to an array of TestMessageDecorations. Only test results that are completed (have completedAt !== undefined) should be cached there, and we should also make sure to remove any cached messages `invalidateResultMessage`.

**Comment by connor4312** [comment]
Are you planning to add caching in this PR, or should we go ahead and merge as-is? 🙂

**Review by connor4312 — APPROVED** [review]

**Review by joyceerhl — APPROVED** [review]

**Pull request merged**